File: src/main/java/codelity/Lesson4.java
package codelity;

import java.util.Arrays;
import java.util.HashMap;

public class Lesson4 {

    public int FrogRiverOne(int[] A, int X) {
        HashMap<Integer, Integer> positionToTimeMap = new HashMap<>();

        for (int time = 0; time < A.length; time++) {
            if (positionToTimeMap.size() == X) {
                break;
            }

            int position = A[time];
            if (position <= X) {
                Integer earliestTime = positionToTimeMap.get(position);
                if (earliestTime != null) {
                    if (time < earliestTime) {
                        positionToTimeMap.put(position, time);
                    }
                } else {
                    positionToTimeMap.put(position, time);
                }
            }
        }

        if (positionToTimeMap.size() < X) {
            return -1;
        }

        final int[] maxTime = {Integer.MIN_VALUE};

        positionToTimeMap.forEach((position, time) -> {
            if (time > maxTime[0]) {
                maxTime[0] = time;
            }
        });

        return maxTime[0];
    }

    public int PermCheck(int[] A) {
        int[] occurrences = new int[A.length + 1];
        for (int i = 0; i < A.length; i++) {
            int number = A[i];
            if (number > A.length) {
                return 0;
            }
            occurrences[number] = occurrences[number] + 1;
        }

        for (int i = 1; i < occurrences.length; i++) {
            if (occurrences[i] != 1) {
                return 0;
            }
        }

        return 1;
    }

    public int[] MaxCounters(int N, int[] A) {
        int maxValue = 0;
        int minValue = 0;
        int[] counters = new int[N];

        for (int i = 0; i < A.length; i++) {
            int operation = A[i];
            if (operation == N + 1) {
                minValue = maxValue;
            } else {
                if (counters[operation-1] < minValue) {
                    counters[operation-1] = minValue + 1;
                } else {
                    counters[operation-1]++;
                }

                int counterValue = counters[operation-1];
                if (maxValue < counterValue) {
                    maxValue = counterValue;
                }
            }
        }

        if (minValue > 0)
            for (int i = 0; i < counters.length; i++) {
                if (counters[i] < minValue) {
                    counters[i] = minValue;
                }
            }

        return counters;
    }


    public int MissingInteger(int[] A){
        Arrays.sort(A);
        int missingInteger = 1;

        for (int i = 0; i < A.length; i++) {
            if (A[i] == missingInteger){
                missingInteger++;
            }else{
                break;
            }
        }

        return missingInteger;
    }
}

File: src/main/js/codelity/lesson4.ts
import * as buffer from "buffer";

function FrogRiverOne(X: number, A: number[]) {
    const existingLeaves = new Set<number>();
    for (let i = 0; i < A.length; i++) {
        const second = i;
        const element: number = A[i];
        if (element <= X) {
            existingLeaves.add(element);
            if (existingLeaves.size === X) {
                return second;
            }
        }
    }

    return -1;
}

// console.log(FrogRiverOne(5, [1, 3, 1, 4, 2, 3, 5, 4]));


function PermCheck(A: number[]) {
    const IS_PERM = 1;
    const IS_NOT_PERM = 0;
    const set = new Set<number>()

    for (const number of A) {
        if (number > A.length) {
            return IS_NOT_PERM;
        }
        if (set.has(number)) {
            return IS_NOT_PERM;
        }
        set.add(number);
    }

    return set.size === A.length ? IS_PERM : IS_NOT_PERM;
}

console.log(PermCheck([4, 1, 3, 2]));


function MaxCounters(N: number, A: number[]) {
    let minValue = 0;
    let maxValue = 0;
    const solution = new Array(N).fill(0);

    for (const counter of A) {
        if (counter > N) {
            minValue = maxValue;
        } else {
            const index = counter - 1;
            let value = solution[index];
            if (value < minValue) {
                value = minValue;
            }
            ++value;
            if (value > maxValue) {
                maxValue = value;
            }
            solution[index] = value;
        }
    }

    return solution.map(e => e < minValue ? minValue : e);
}

console.log(MaxCounters(5, [3, 4, 4, 6, 1, 4, 4]))

function MissingInteger(A: number[]) {
    const N = A.length;
    const present = new Set<number>(); // Track present numbers

    for (const num of A) {
        if (num > 0 && num <= N) {
            present.add(num);
        }
    }

    for (let i = 1; i <= N; i++) {
        if (!present.has(i)) {
            return i;
        }
    }

    return N + 1;
}

console.log(MissingInteger([-1, -3]));

